src/DIRAC/DataManagementSystem/Agent/RequestOperations/test/Test_ReplicateAndRegister.py

```diff
@@ -0,0 +1,690 @@
+"""
+import errno
+import json
+import os
+import re
+import pytest
+import DIRAC
+
+from abc import ABC, abstractmethod
+
+from zlib import adler32
+
+from DIRAC import S_OK, S_ERROR
+from DIRAC.DataManagementSystem.Agent.RequestOperations.ReplicateAndRegister import filterReplicas
+
+from DIRAC.RequestManagementSystem.Client.File import File
+
+
+# Although we will in general use regular expression to drive the behavior
+# of the mocks, there are a few very specific files.
+# So we may as well just use them as constant.
+# They are mostly cases that return S_ERROR straight away....
+
+# LFN for a non existing file
+NON_EXISTING_LFN = "/lhcb/i/dont/exist.txt"
+
+# Complete failure of getting replicas
+ERROR_GETTING_REPLICAS = "/lhcb/S_ERROR/replicas"
+
+# The file simply has no replicas available
+NO_AVAILABLE_REPLICAS = "/lhcb/no/available/replicas"
+
+
+# Complete failure of getting FC metadata
+ERROR_GETTING_FC_METADATA = "/lhcb/S_ERROR/fc_metadata"
+
+# Complete failure of getting SE metadata
+ERROR_GETTING_SE_METADATA = "/lhcb/S_ERROR/se_metadata"
+
+# SE at which we will find the files by default
+DEFAULT_SE = "DiskSE"
+
+
+# Now starts the regular expression magic.
+# The idea is that the LFN will be structured like
+# /MARKER_XXX/value1/value2/END/MARKER_YYY/value3/END...
+# We split the LFN based on the marker and end mark,
+# and each mock looks for the marker it cares about
+# (for example MARKER_BAD_SE_REPLICAS for the mock_SE_getReplicas)
+
+# The file replicas at given SEs
+# The "Disk" and "Tape" part are interpreted as such
+MARKER_WITH_REPLICAS = "WITH_REPLICAS"
+
+# To declare that the specific replica has wrong cks
+MARKER_BAD_SE_REPLICAS = "BAD_SE_REPLICAS"
+
+# The cks is wrong in the FC
+MARKER_BAD_FC_REPLICAS = "BAD_FC_REPLICAS"
+
+# The file physicaly does not exist but is registered
+MARKER_BAD_SE_NOT_EXIST = "BAD_SE_NOT_EXIST"
+
+# The end marker, same for all
+END_MARKER = "END"
+
+# Collect all the markers to build the regular expression
+ALL_MARKERS = "|".join(v for k, v in globals().items() if k.startswith("MARKER_"))
+# That is the regular expression to split the LFN
+# /MARKER1/val1/val2/END/MARKER2/val3/END...
+MARKER_PATTERN = rf"/*(?P<pat>({ALL_MARKERS})+)/(?P<values>.*)/"
+
+
+def _splitLFN(lfn):
+    """Split the LFN based on the pattern"""
+    instructions = {}
+    try:
+        # Split at each end marker
+        for group in re.split(rf"{END_MARKER}", lfn):
+            if not group:
+                continue
+
+            # interpret the specific pattern
+            matchgroup = re.match(MARKER_PATTERN, group).groupdict()
+            instructions[matchgroup["pat"]] = matchgroup["values"]
+    except Exception:
+        pass
+    return instructions
+
+
+def mock_DM_getReplicas(self, lfns, **kwargs):
+    """Mock the DataManager.getReplicas method.
+
+    It returns based on the LFN.
+    See the code for details.
+    :param preferDisk: only return disk replicas if any
```

Comment:
```suggestion

    :param preferDisk: only return disk replicas if any
```

Not that we sphinx the test function, but out of principle